### ISSUE TITLE
bug: Cannot implement generic trait for component

### ISSUE THREAD

**Issue by enitrat** [starkware-libs/cairo#4724]
# Bug Report

**Cairo version:**
2.3.1

**Current behavior:**
Here is the trait that I want my components to share. Each component represents an exchange, and each component implements it to manage adding LP to a pool of the specific exchange the component is implemented for

```
trait IAmmAdapter<TContractState> {
    fn create_and_add_liquidity(
        ref self: TContractState,
        exchange_address: ContractAddress,
        token_address: ContractAddress,
        counterparty_address: ContractAddress,
        unlock_time: u64,
        additional_parameters: Span<felt252>,
    ) -> Span<felt252>;
}

```

As you can the function returns a Span<felt252> - this is because I cannot implement this trait as generic over a return type `R`, like this:

```rust
trait IAmmAdapter<TContractState, R> {
    fn create_and_add_liquidity(
        ref self: TContractState,
        exchange_address: ContractAddress,
        token_address: ContractAddress,
        counterparty_address: ContractAddress,
        unlock_time: u64,
        additional_parameters: Span<felt252>,
    ) -> R;
}
```
If I define the trait like this and try to implement it for a specific R:u64 type as follows:
```rust
 #[embeddable_as(EkuboAdapterImpl)]
    impl EkuboAdapter<
        TContractState,
        +HasComponent<TContractState>,
       //... a bunch of other requirements
    > of unruggable::exchanges::IAmmAdapter<ComponentState<TContractState>, u64> {
        fn create_and_add_liquidity(
          // snip
        ) -> u64 {
```

Then, I get the error `Plugin diagnostic: Type annotations needed. Failed to infer ?0`

If I dive a bit deeper and check what the `EkuboAdapterImpl` generated impl looks like:

```rust
```

**Expected behavior:**

**Steps to reproduce:**

**Related code:**

```
insert short code snippets here
```

**Other information:**

**Comment by orizi** [comment]
Have you marked `IAmmAdapter` as `#[starknet::interface]`?

**Comment by enitrat** [comment]
> Have you marked IAmmAdapter as #[starknet::interface]?

No, I haven't - and I don't see why I would 🤔 I don't want it to be an interface, I want it to be some kind of internal trait. It's never exposed to the outside world

**Comment by orizi** [comment]
so you don't need to have it embeddable at all.
embeddable_as is to prepare an impl in a component to be `#[abi(embed_v0)]` in a contract.

Remove the `embeddable_as` and just remove use it as an impl for a contract implementing the 'HasComponent'.

Will add a diagnostic for `embeddable` not a starknet interface.

**Comment by enitrat** [comment]
> embeddable_as is to prepare an impl in a component to be #[abi(embed_v0)] in a contract

ah, ok, it wasn't clear to me that a component impl could be brought to a contract without being embeddable_as.

Thanks!
